src/batch/src/executor/source.rs

```diff
@@ -12,19 +12,176 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use risingwave_common::error::Result;
+use std::collections::HashMap;
+use std::sync::Arc;
 
+use futures::StreamExt;
+use futures_async_stream::try_stream;
+use risingwave_common::array::DataChunk;
+use risingwave_common::catalog::{ColumnDesc, ColumnId, Field, Schema, TableId};
+use risingwave_common::error::ErrorCode::{ConnectorError, ProtocolError};
+use risingwave_common::error::{Result, RwError};
+use risingwave_common::types::DataType;
+use risingwave_connector::source::{ConnectorProperties, SplitImpl, SplitMetaData};
+use risingwave_pb::batch_plan::plan_node::NodeBody;
+use risingwave_pb::catalog::source_info::SourceInfo as ProstSourceInfo;
+use risingwave_pb::plan_common::RowFormatType;
+use risingwave_source::connector_source::{ConnectorSource, SourceContext};
+use risingwave_source::monitor::SourceMetrics;
+use risingwave_source::{SourceColumnDesc, SourceFormat, SourceParserImpl};
+
+use super::Executor;
 use crate::executor::{BoxedExecutor, BoxedExecutorBuilder, ExecutorBuilder};
 use crate::task::BatchTaskContext;
 
-pub struct SourceExecutor {}
+pub struct SourceExecutor {
+    connector_source: ConnectorSource,
+
+    // used to create reader
+    column_ids: Vec<ColumnId>,
+    metrics: Arc<SourceMetrics>,
+    source_id: TableId,
+    split: SplitImpl,
+
+    schema: Schema,
+    identity: String,
+}
 
 #[async_trait::async_trait]
 impl BoxedExecutorBuilder for SourceExecutor {
     async fn new_boxed_executor<C: BatchTaskContext>(
-        _source: &ExecutorBuilder<'_, C>,
-        _inputs: Vec<BoxedExecutor>,
+        source: &ExecutorBuilder<'_, C>,
+        inputs: Vec<BoxedExecutor>,
     ) -> Result<BoxedExecutor> {
-        todo!("Can't support SourceExecutor now!")
+        ensure!(inputs.is_empty(), "Source should not have input executor!");
```

Comment:
suggest assert_debug!

Reply:
ensure! here is used to return an error rather than panic. And seems RowScan also use ensure!🤔

https://github.com/risingwavelabs/risingwave/blob/845c5a5523af8b344d3d99df5d5fb9f05df759b4/src/batch/src/executor/row_seq_scan.rs#L159